### PR TITLE
Unsafe job for aws mem realloc

### PR DESCRIPTION
**Pull request by LinerSu**
This PR includes a simple unsafe test for a bug we found on the `aws_mem_realloc` function: 
https://github.com/awslabs/aws-c-common/blob/5fa4275653600b66cdbfee8fcf407eb4416f6b18/source/allocator.c#L217

- Bug descriptions.
`aws_mem_realloc` could failed on calling `mem_cpy`:
https://github.com/awslabs/aws-c-common/blob/5fa4275653600b66cdbfee8fcf407eb4416f6b18/source/allocator.c#L245
Because the `*ptr` could be a `NULL` and the `oldsize > 0`.
- Why this is a bug?
Since there are no preconditions for the `void **ptr` and size `oldsize`. The `*ptr` could be a `NULL` and the `oldsize > 0`.
Another precondition is when there is no mem realloc in the aws allocator (realloc is an optional):
https://github.com/awslabs/aws-c-common/blob/5fa4275653600b66cdbfee8fcf407eb4416f6b18/include/aws/common/allocator.h#L18-L19

### CODE REVIEW

File: seahorn/lib/proof_allocators.c [hunk 1]
```diff
@@ -40,17 +40,49 @@ static void s_free_allocator(struct aws_allocator *allocator, void *ptr) {
   free(ptr);
 }
 
+static void *s_realloc_allocator(struct aws_allocator *allocator, void *ptr, size_t oldsize, size_t newsize) {
```
Comment (priyasiddharth) [inline seahorn/lib/proof_allocators.c:43]: can we use this realloc function

https://sourcegraph.com/github.com/seahorn/verify-c-common/-/blob/seahorn/lib/sea_allocators.c?L72&subtree=true

File: seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#include "nondet.h"
+
+int main() {
+
+    struct aws_allocator *allocator = nd_bool() ? sea_allocator() : sea_allocator_with_realloc();
+
+    void *data = NULL;
+
+    size_t old_size = nd_size_t();
+    assume(old_size <= UINT8_MAX);
```
Comment (priyasiddharth) [inline seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c:17]: is this assume really needed?

**Review by priyasiddharth — COMMENTED** [review]
I reviewed just because it was open :) please feel free to ignore

File: seahorn/lib/proof_allocators.c (hunk 1 above)
Comment (LinerSu) [reply seahorn/lib/proof_allocators.c:43]: I do not get your point. The `s_realloc_allocator` has a different function type from `sea_realloc`.  Do you mean this:
We change
```c
void *new_mem = s_malloc_allocator(allocator, newsize);
```
into
```c
void *new_mem = sea_realloc(ptr, newsize);
```

File: seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c (hunk 2 above)
Comment (LinerSu) [reply seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c:17]: Not necessary. The size could be unbounded.

File: seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    size_t new_size = nd_size_t();
+    assume(old_size <= new_size);
+
+    aws_mem_realloc(allocator, (void **)&data, old_size, new_size);
+
+    sassert(sea_is_dereferenceable(data, new_size));
```
[the remark this answers is not in the document]
Comment (LinerSu) [reply seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c:24]: Yes. We can delete this one.

File: seahorn/lib/proof_allocators.c [hunk 4]
```diff
@@ -40,17 +40,49 @@ static void s_free_allocator(struct aws_allocator *allocator, void *ptr) {
...
+  (void)oldsize;
+  AWS_FATAL_PRECONDITION(newsize);
+
+  if (newsize <= oldsize) {
+    return ptr;
+  }
+
+  /* newsize is > oldsize, need more memory */
+  void *new_mem = s_malloc_allocator(allocator, newsize);
+  // AWS_PANIC_OOM(new_mem, "Unhandled OOM encountered in s_malloc_allocator");
```
Comment (priyasiddharth) [inline seahorn/lib/proof_allocators.c:54]: Can realloc return a `NULL` ptr? If so then let's just remove the commented line.

**Review by priyasiddharth — CHANGES_REQUESTED** [review]

File: seahorn/lib/proof_allocators.c (hunk 4 above)
Comment (LinerSu) [reply seahorn/lib/proof_allocators.c:54]: Yes. `realloc` can fail and return a null pointer.

File: seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+int main() {
+
+    struct aws_allocator *allocator = nd_bool() ? sea_allocator() : sea_allocator_with_realloc();
+
+    void *data = NULL;
+
+    size_t old_size = nd_size_t();
+
+    size_t new_size = nd_size_t();
+    assume(old_size <= new_size);
```
[the remark this answers is not in the document]
Comment (LinerSu) [reply seahorn/jobs_unsafe/mem_realloc_unsafe/aws_mem_realloc_unsafe_harness.c:19]: No need. Removed.

File: seahorn/jobs_unsafe/mem_realloc_unsafe/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+add_executable(mem_realloc_unsafe
+  ${AWS_C_COMMON_ROOT}/source/allocator.c
+  ${SEA_LIB}/bcmp.c
+  ${SEA_LIB}/proof_allocators.c
+  ${SEA_LIB}/sea_allocators.c
+  aws_mem_realloc_unsafe_harness.c)
+
+sea_link_libraries(mem_realloc_unsafe sea_bounds.ir)
+sea_attach_bc(mem_realloc_unsafe)
+sea_add_sat_test(mem_realloc_unsafe)
```
Comment (priyasiddharth) [inline seahorn/jobs_unsafe/mem_realloc_unsafe/CMakeLists.txt:10]: Please add justification for why this is `sat`.

**Review by priyasiddharth — APPROVED** [review]
LGTM except minor comment

**Pull request merged**